Adds Teleprods

PKPenguin321 opened the pull request:
Crafted by sticking a bluespace crystal (artificial or real) onto a power cell-less stun prod, or by tablecrafting with the ingredients for a stunprod + a bluespace crystal (again, it can be artificial).

They are functionally similar to stunprods, requiring a power cell and having a limited number of charges.

When somebody is stunned by it, they are also teleported away to a random nearby area. This _can_ space them/put them into walls. The teleport works even if the stun doesn't (IE, if you hit a xeno it would teleport but they would not get stunned).

![teleprod sprites](https://cloud.githubusercontent.com/assets/8345184/11414362/60b3c638-93a9-11e5-874d-8502cc20d52e.png)

:cl: PKPenguin321
rscadd: Teleprods can now be constructed. Make them via tablecrafting, or by putting a bluespace crystal onto a stunprod. They warp the target to a random nearby area (which may include space/the inside of a wall) in addition to giving the victim a brief stun. Like stunprods, they require a power cell to function.
rscadd: Teleprods can have their crystal removed and be made back into stunprods by using them in your hand when they have no cell installed.
/:cl:

PKPenguin321 committed:
adds teleprods, a way to make them with bluespace crystals and a stunprod, and a tablecrafting recipe

PKPenguin321 commented:
aaaand i forgot to make them use the new inhands

optimumtact commented on tgstation.dme:
```diff
@@ -922,7 +923,6 @@
 #include "code\modules\awaymissions\mission_code\spacebattle.dm"
 #include "code\modules\awaymissions\mission_code\stationCollision.dm"
 #include "code\modules\awaymissions\mission_code\wildwest.dm"
-#include "code\modules\client\asset_cache.dm"
```
h-h-hey what are you d-d-doing

PKPenguin321 replied:
shhhh

JJRcop replied:
This happens a lot, it's very likely he didn't mean to do it.

Zilenan91 commented:
Can you remove the stun on it?

PKPenguin321 commented:
>Can you remove the stun on it?

i mean i could but why would i

PKPenguin321 committed:
fixes tgstation.dme since i cucked it by unticking a file accidentally, also makes the teleprods use the correct inhands

Zilenan91 commented:
Because it doesn't need or make sense to have a stun on it. Bluespace crystals don't really stun people, and the teleportation combined with the stun can fuck people over more than just being prodded does

PKPenguin321 commented:
it's a weapon
it fucking people over is kind of the idea ¯\\\_(ツ)_/¯

Zilenan91 commented:
Then make a stunprod if you want to stun people. The stun combined with the teleportation is too much.

MMMiracles commented:
zilenan stop hating fun.

people are totally gonna be raiding the station teleporter more often now solely for those sweet, sweet crystals though.

Incoming5643 commented:
Stick with the prod

GunHog commented:
Do not be lazy! Allow this to be done with proper Security batons as well.

KorPhaeron commented:
Sounds pretty useless but fun

optimumtact commented:
>gunhog wants a subtle security buff
COLOUR ME SUPRISED

PKPenguin321 commented:
>Do not be lazy! Allow this to be done with proper Security batons as well.

this would basically double the size of this pr, so nah
plus how would you even mount a bluespace crystal to a baton i mean cmon

GunHog commented:
I do not play Security, Oranges - stop confusing me with Steelpoint. It did not make sense to me for the child item to have it, but not the parent. 

You would mount the crystal the same way you did for your piece of rod, wirecutters, and cable - by stuffing it between the terminals. If your super delicate crystal held in place with a sharp, metallic clamp which itself is tied to a metal rod with some uninsulated wire makes sense, then so would that. It does not really matter to me as a player, as I can much more easily access a stun prod than I could a baton.

PKPenguin321 commented:
i guess but again that would require a shitload more code and prods work fine anyways

WJohn commented:
If it DOES stun them, I think it should do the stunning *after* it teleports them, that way you don't have them just drop the items they were holding and be too far to get them back. It'd be an intensely powerful inhand-stealing tool.

KorPhaeron commented:
It requires a rare mining material+you give up the chance to kill reliably it's fine as is

Zilenan91 commented:
Or just decon one of the teleporters like the one on the AI sat.

PKPenguin321 commented:
balance police please go

GunHog commented:
Oh god, the filthy greytide is going to breaking into the teleporter room each round. I did not even think about this.

PKPenguin321 commented:
maybe for like three rounds before the novelty wears off

GunHog commented:
When I play RD, I should consider building a turret or two in there to discourage theft...

Zilenan91 commented:
do it and set it to laser. Cut the AI camera too so it doesn't deactivate it.

Lumipharon commented:
So you'll be able to teleport yourself with this I assume (even if it stuns you if you're not hulk).
This seems fairly strong, and basically just incentivises ripping machines up to make weapons.

GunHog commented:
Using the weapon on oneself makes it one-use for most except hulks and anti-drop implant users, as the stun happens first, causing the user to drop the weapon.

Still, it is a valid concern, even if the user takes the risk of stunning himself and losing the weapon first.

PKPenguin321 commented:
if anybody has a better idea for implementing these that doesnt require bluespace crystals, hit me with it either here or on irc
for now im just gonna leave it like this since i think you guys are really *really* **really** overestimating the balance aspects of this

Dorsisdwarf commented:
This looks like everyone's overthought it so much

If you catch an assistant with one of these as sec, insta-perma them for tresspass, theft, and posession of a weapon. There's basically no way to legitimately have one of these. boom.

Lumipharon commented:
printing a crystal with R&D, making one at xeno, or finding one at mining are all genuine ways to get bluespace crystals. Also looting off station teleporters.

Stenthus commented:
how far away does it teleport people?

Dorsisdwarf commented:
:s HOS ask the RD if they're handing out telecrystals to greyshirts

dumpdavidson commented:
>Then make a stunprod if you want to stun people. The stun combined with the teleportation is too much.

But teleporting someone you just stunned away is worse than just stunning them. 

I agree with Kor here: useless but fun.

Alek2ander commented:
You know you can throw the crystals, right?

CosmicScientist commented:
@Alek2ander that uses up the crystal, from the looks of this you get infinite uses for tapping people with a rod, the only drawback is replacing/recharging the cell every-so-often (I assume as much as the regular stunprod).

optimumtact commented:
we can always add crystal degradation if it proves to be too annoying.

PKPenguin321 commented:
@CosmicScientist Yep
@Stenthus It should warp between 1 and 15 tiles away. If a clown accidentally hits themselves with it out of clumsiness, they can warp up to 50 tiles away (honk)

PKPenguin321 commented:
Do not merge for now, mining changes got merged and changed the pathing of bluespace crystals, meaning this won't work until I update it
Also I'm gonna try to add the ability to remove the crystal

PKPenguin321 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into teleprods

duncathan commented:
I restarted your Travis build just for you @PKPenguin321

PKPenguin321 committed:
fixes conflicts, adds the ability to remove the crystal from the prod via using it when no cell is attached

PKPenguin321 commented:
So yeah, conflicts fixed, also you can now remove the crystal from a teleprod and make it back into a stunprod by using it in your hand when it has no cell installed. @GunHog I also found a way to make batons work, too, but I'm gonna save it for a separate PR to avoid the monster of controversy that is additions security.

Bawhoppen commented:
would be top-tier to have one that teleports every item on your character to a different random location


also should be able to teleport other crap you hit like chairs and shit

PKPenguin321 commented:
>would be top-tier to have one that teleports every item on your character to a different random location

would be hilarious, but not happening for both code and balance reasons

>also should be able to teleport other crap you hit like chairs and shit

nah, for the same reasons (plus "Who stole the particle accelerator?!" "Issue Report: Singularity can be warped away with a teleprod")

KorPhaeron commented:
@PKPenguin321 

Could just make it check for anchored, which I'm hoping it already does for humans

PKPenguin321 commented:
Yeah, they do check for anchored on humans

This is also pretty much ready for review/merge

duncathan commented on code/game/objects/items/weapons/teleprod.dm:
```diff
@@ -0,0 +1,50 @@
+		return
+	else
+		if(status)
+			if(!istype(M) && M.anchored)
```
Are you sure this should be `&&`?

PKPenguin321 replied:
honestly i have no clue, i was hoping reviewers could tell me
i can do ingame tests real fast, hold on

GunHog replied:
Testing is fun :)

PKPenguin321 replied:
okay so after fixing teleprods and testing again, i can confirm the anchor check works. if a mob is buckled (anchored) and hit with a teleprod, they automatically unbuckle, meaning there's no problemo.

the only real issue i found was that if you're clumsy and accidentally hit yourself when trying to hit someone else, they still teleport even though they're not stunned. not sure how to fix this one.

PKPenguin321 committed:
fixes everything
i broke everything by giving teleprods a unique attack_self but forgetting to include the original functions of attack_self, making them unable to turn on/off. this fixes that.

PKPenguin321 commented:
WOW @Cheridan 
RIGHT AS I MAKE MY COMMIT TO FIX THE BUGS YOU MERGE MY OTHER PR WHICH MAKES THE INHANDS CONFLICT
REEEEEEEEEEEEEEEEEEEEEEEEEEEEEE

PKPenguin321 committed:
fixes the conflicts
....by updating the inhand files. reeeeeeeeeeeee

PKPenguin321 committed:
Merged upstream

PKPenguin321 committed:
merged upstream ps fuck git

PKPenguin321 commented:
fucking updating the inhands was harder than the actual coding, fucking git :rage:
